examples/common/pigweed/rpc_services/AttributeAccessor.h

```diff
@@ -0,0 +1,101 @@
+/*
+ *
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "pw_status/status.h"
+#include <app/AttributeReportBuilder.h>
+#include <app/AttributeValueDecoder.h>
+#include <app/AttributeValueEncoder.h>
+
+/**
+ * Callback class that clusters can implement in order to interpose custom
+ * attribute-handling logic.  An AttributeAccessor instance is associated
+ * with some specific cluster.  A single instance may be used for a specific
+ * endpoint or for all endpoints.
+ *
+ * Instances of AttributeAccessor that are registered via
+ * AttributeAccessorRegistry::Instance().Register will be consulted before taking the
+ * normal attribute access codepath and can use that codepath as a fallback if desired.
+ */
+namespace chip {
+namespace rpc {
+
+class AttributeAccessor
+{
+public:
+    /**
+     * aEndpointId can be Missing to indicate that this object is meant to be
+     * used with all endpoints.
+     */
+    AttributeAccessor(Optional<EndpointId> aEndpointId, ClusterId aClusterId) : mEndpointId(aEndpointId), mClusterId(aClusterId) {}
+    virtual ~AttributeAccessor() {}
+
+    /**
+     * Callback for reading attributes.
+     *
+     * @param [in] aPath indicates which exact data is being read.
+     * @param [in] aEncoder the AttributeValueEncoder to use for encoding the
+     *             data.
+     *
+     * The implementation can do one of three things:
+     *
+     * 1) Return a failure.  This is treated as a failed read and the error is
+     *    returned to the client, by converting it to a StatusIB.
+     * 2) Return success and attempt to encode data using aEncoder.  The data is
+     *    returned to the client.
+     * 3) Return success and not attempt to encode any data using aEncoder.  In
+     *    this case, Ember attribute access will happen for the read. This may
+     *    involve reading from the attribute store or external attribute
+     *    callbacks.
+     */
+    virtual ::pw::Status Read(const chip::app::ConcreteReadAttributePath & aPath, chip::app::AttributeValueEncoder & aEncoder) = 0;
```

Comment:
Suggest default implementation, since reads are unlikely to be overridden, so to avoid forcing users to handle read (which can cleanly go through prior path), suggest having a default impl.